File: Chapter 02/DecoratorDesignPatternExample.py
class Cake:
    def get_description(self):
        pass
    def cost(self):
        pass
class Sponge(Cake):
    def get_description(self):
        return "Sponge"
    def cost(self):
        return 2.0
class CakeDecorator(Cake):
    def __init__(self, decorated_cake):
        self.decorated_cake = decorated_cake
    def get_description(self):
        return self.decorated_cake.get_description()
    def cost(self):
        return self.decorated_cake.cost()
class FrostingDecorator(CakeDecorator):
    def get_description(self):
        return super().get_description() + ", with Frosting"
    def cost(self):
        return super().cost() + 0.5
class SprinklesDecorator(CakeDecorator):
    def get_description(self):
        return super().get_description() + ", with Sprinkles"
    def cost(self):
        return super().cost() + 0.2
if __name__ == "__main__":
    sponge = Sponge()
    print("Cost of Sponge Cake: $" + str(sponge.cost()))
    frosting_coffee = FrostingDecorator(sponge)
    print("Cost of Cake with Frosting: $" + str(frosting_cake.cost()))
    sprinkles_frosting_cake = SprinklesDecorator(frosting_cake)
    print("Cost of Cake with Frosting and Sprinkles: $" + str(sprinkles_frosting_cake.cost()))

File: Chapter 02/DecoratorDesignPatternExample.java
interface Cake {
    String getDescription();
    double cost();
}


class Sponge implements Cake {
    public String getDescription() {
        return "Sponge";
    }
    public double cost() {
        return 2.0;
    }
}

abstract class CakeDecorator implements Cake {
    protected Cake decoratedCake;
    public CakeDecorator(Cake decoratedCake) {
        this.decoratedCake = decoratedCake;
    }
    public String getDescription() {
        return decoratedCake.getDescription();
    }
    public double cost() {
        return decoratedCake.cost();
    }
}

class FrostingDecorator extends CakeDecorator {
    public FrostingDecorator(Cake decoratedCake) {
        super(decoratedCake);
    }
    public String getDescription() {
        return super.getDescription() + ", with Frosting";
    }
    public double cost() {
        return super.cost() + 0.5;
    }
}
class SprinklesDecorator extends CakeDecorator {
    public SprinklesDecorator(Cake decoratedCake) {
        super(decoratedCake);
    }
    public String getDescription() {
        return super.getDescription() + ", with Sprinkles";
    }
    public double cost() {
        return super.cost() + 0.2;
    }
}


public class DecoratorDesignPatternExample {
    public static void main(String[] args) {
        Cake sponge = new Sponge();
        System.out.println("Cost of Sponge Cake: $" + sponge.cost());

        Cake frostingCake = new FrostingDecorator(sponge);
        System.out.println("Cost of Cake with Frosting : $" + frostingCake.cost());

        Cake sprinklsSpongeCake = new SprinklesDecorator(frostingCake);
        System.out.println("Cost of Cake with Frosting and Sprinkles: $" + sprinklesFrostingCake.cost());
    }
}

